provider/datagen/src/transform/segmenter/mod.rs

```diff
@@ -780,6 +783,135 @@ impl IterableResourceProvider<UCharDictionaryBreakDataV1Marker> for SegmenterDic
     }
 }
 
+// ndarray data structure in LSTM JSON data.
+#[derive(serde::Deserialize, Debug)]
+struct RawLstmMatrix {
+    #[allow(dead_code)]
+    pub v: i16,
+    pub dim: Vec<i16>,
+    pub data: Vec<f32>,
+}
+
+impl RawLstmMatrix {
+    fn as_lstm_matrix(&self) -> LstmMatrix<'static> {
+        LstmMatrix {
+            dim: ZeroVec::alloc_from_slice(&self.dim),
+            data: ZeroVec::alloc_from_slice(&self.data),
+        }
+    }
+}
+
+// LSTM JSON data structure.
+#[derive(serde::Deserialize, Debug)]
+struct RawLstmData {
+    model: String,
+    dic: HashMap<String, i16>,
+    mat1: RawLstmMatrix,
+    mat2: RawLstmMatrix,
+    mat3: RawLstmMatrix,
+    mat4: RawLstmMatrix,
+    mat5: RawLstmMatrix,
+    mat6: RawLstmMatrix,
+    mat7: RawLstmMatrix,
+    mat8: RawLstmMatrix,
+    mat9: RawLstmMatrix,
+}
+
+/// A data provider reading from segmenter lstm files.
+#[derive(Debug)]
+pub struct SegmenterLstmProvider {
+    source: SourceData,
+}
+
+impl From<&SourceData> for SegmenterLstmProvider {
+    fn from(source: &SourceData) -> Self {
+        Self {
+            source: source.clone(),
+        }
+    }
+}
+
+impl SegmenterLstmProvider {
+    // Generate LSTM Data for DataProvider from LSTM JSON.
+    fn generate_data(&self, options: &ResourceOptions) -> Result<LstmDataV1<'static>, DataError> {
+        let lstm_data: &RawLstmData = self
+            .source
+            .segmenter_lstm()?
+            .read_and_parse::<RawLstmData>(
+                Self::get_json_filename(options)
+                    .ok_or_else(|| DataErrorKind::MissingResourceOptions.into_error())?,
+            )?;
+        // The map of "dic" may not be sorted, so we cannot use ZeroMap directly.
+        let mut dic: ZeroMap<'static, str, i16> = ZeroMap::new();
```

Comment:
Question: Why is it a map from `str` instead of `char`? Is it only to support grapheme mode?

Reply:
When adding grapheme mode, we have to modify data model. So I use `str` instad of `char`.